examples/gno.land/r/gnoland/valopers/valopers.gno

```diff
@@ -5,35 +5,173 @@ package valopers
+	}
+
+	// Remove the old valoper info
+	// in case the address changed
+	if address != v.Address {
+		valopers.Remove(address.String())
+	}
+
+	// Save the new valoper info
+	valopers.Set(v.Address.String(), v)
+}
+
+// GetByAddr fetches the valoper using the address, if present
+func GetByAddr(address std.Address) Valoper {
+	valoperRaw, exists := valopers.Get(address.String())
+	if !exists {
+		panic(errValoperMissing)
+	}
+
+	return valoperRaw.(Valoper)
+}
 
-	DisplayName   string      // The display name of the valoper.
-	ValidatorAddr std.Address // The address of the validator.
-	// TODO: Add other valoper metadata as needed.
+// Render renders the current valoper set
+func Render(_ string) string {
+	if valopers.Size() == 0 {
+		return "No valopers to display."
+	}
+
+	output := "Valset changes to apply:\n"
+	valopers.Iterate("", "", func(_ string, value interface{}) bool {
+		valoper := value.(Valoper)
+
+		output += valoper.Render()
+
+		return false
+	})
+
+	return output
 }
 
-// Register registers a new valoper.
-// TODO: Define the parameters and implement the function.
-func Register( /* TBD */ ) {
-	panic("not implemented")
+// Render renders a single valoper with their information
+func (v Valoper) Render() string {
+	output := ufmt.Sprintf("## %s\n", v.Name)
+	output += ufmt.Sprintf("%s\n\n", v.Description)
+	output += ufmt.Sprintf("- Address: %s\n", v.Address.String())
+	output += ufmt.Sprintf("- PubKey: %s\n", v.PubKey)
+	output += "- P2P Addresses: [\n"
+
+	if len(v.P2PAddresses) == 0 {
+		output += "]\n"
+
+		return output
+	}
+
+	for index, addr := range v.P2PAddresses {
+		output += addr
+
+		if index == len(v.P2PAddresses)-1 {
+			output += "]\n"
+
+			continue
+		}
+
+		output += ",\n"
+	}
+
+	return output
 }
 
-// Update updates an existing valoper.
-// TODO: Define the parameters and implement the function.
-func Update( /* TBD */ ) {
-	panic("not implemented")
+// isValoper checks if the valoper exists
+func isValoper(address std.Address) bool {
+	_, exists := valopers.Get(address.String())
+
+	return exists
 }
 
-// GovXXX is a placeholder for a function to interact with the governance DAO.
-// TODO: Define a good API and implement it.
-func GovXXX() {
-	// Assert that the caller is a member of the governance DAO.
-	govdao.AssertIsMember(std.PrevRealm().Addr())
-	panic("not implemented")
+// GovDAOProposal creates a proposal to the GovDAO
+// for adding the given valoper to the validator set.
+// This function is meant to serve as a helper
+// for generating the govdao proposal
+func GovDAOProposal(address std.Address) {
+	valoper := GetByAddr(address)
+	if !valoper.Active {
+		panic(errValoperInactive)
+	}
+
+	// Make sure the valoper is the caller
+	if std.GetOrigCaller() != address {
+		panic(errValoperInactive)
+	}
+
+	changesFn := func() []pVals.Validator {
+		return []pVals.Validator{
+			{
+				Address:     valoper.Address,
+				PubKey:      valoper.PubKey,
+				VotingPower: 10,
```

Comment:
```suggestion
				VotingPower: 1,
```

Let's start with this approach and consider making it configurable in a more advanced version later.

Reply:
```suggestion
				VotingPower: int(valoper.Active),
```

Or like this?

Reply:
I've updated it based on the `valoper.Active` state:
[c54e70b](https://github.com/gnolang/gno/pull/2509/commits/c54e70b3d3f6f5128d33651e0c3a82f690392eb3)